src/hotspot/share/oops/compressedKlass.cpp

```diff
@@ -25,46 +25,159 @@
 #include "precompiled.hpp"
 #include "logging/log.hpp"
 #include "memory/metaspace.hpp"
-#include "oops/compressedKlass.hpp"
+#include "oops/klass.hpp"
+#include "oops/compressedKlass.inline.hpp"
 #include "runtime/globals.hpp"
+#include "runtime/java.hpp"
 #include "runtime/os.hpp"
 #include "utilities/debug.hpp"
 #include "utilities/globalDefinitions.hpp"
 #include "utilities/ostream.hpp"
 
-address CompressedKlassPointers::_base = nullptr;
-int CompressedKlassPointers::_shift = 0;
-size_t CompressedKlassPointers::_range = 0;
+int CompressedKlassPointers::_tiny_cp = -1;
+int CompressedKlassPointers::_narrow_klass_pointer_bits = -1;
+int CompressedKlassPointers::_max_shift = -1;
+#ifdef ASSERT
+address CompressedKlassPointers::_klass_range_start = (address)-1;
+address CompressedKlassPointers::_klass_range_end = (address)-1;
+narrowKlass CompressedKlassPointers::_lowest_valid_narrow_klass_id = (narrowKlass)-1;
+narrowKlass CompressedKlassPointers::_highest_valid_narrow_klass_id = (narrowKlass)-1;
+#endif
+
+address CompressedKlassPointers::_base = (address)-1;
+int CompressedKlassPointers::_shift = -1;
+size_t CompressedKlassPointers::_range = (size_t)-1;
```

Comment:
Doesn't this put 0x00000000FFFFFFFF in this size_t? I guess this works with the code using it, but it is not obvious that this is what we intended to put into the _range variable as the (uninitialized) value.

Related to this. How important is it to do these initialized checks? It seems to add casts and a bit odd code. When this is stable, wouldn't it be nice to clean this up. That would also make _tiny_cp a bool instead of a tri-bool.

Reply:
> Doesn't this put 0x00000000FFFFFFFF in this size_t?

Ignore that. I checked and this is signed extended out to 0xFFFFFFFFFFFFFFFF.